src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -2352,51 +2434,105 @@ private void ReportNullabilityMismatchWithTargetDelegate(SyntaxNode syntax, Name
+                    // cf. Binder.CreateUserDefinedConversion
                     {
-                        if (conversion.BestUserDefinedConversionAnalysis.Kind == UserDefinedConversionAnalysisKind.ApplicableInLiftedForm)
+                        if (!conversion.IsValid)
                         {
-                            // PROTOTYPE(NullableReferenceTypes): Conversions: Lifted operator
+                            canConvert = false;
                             break;
                         }
-                        // PROTOTYPE(NullableReferenceTypes): Update method based on operandType.
-                        var resultType = methodOpt.ReturnType;
-                        // Ensure converted type matches expected target type, specifically for lifted
-                        // conversions. See LocalRewriter.MakeConversionNode for similar handling
-                        // based on conversion.BestUserDefinedConversionAnalysis.ToType.
-                        if (!fromConversionNode)
-                        {
-                            if (resultType.TypeSymbol != targetType)
-                            {
-                                // PROTOTYPE(NullableReferenceTypes): Should handle nested nullability
-                                // and top-level nullability. Conversion from resultType to targetType
-                                // should be treated as a built-in conversion.
-                                resultType = TypeSymbolWithAnnotations.Create(targetType);
-                            }
-                        }
-                        return resultType;
+
+                        // operand -> conversion "from" type
+                        operandType = ApplyConversion(
+                            node,
+                            operandOpt,
+                            conversion.UserDefinedFromConversion,
+                            conversion.BestUserDefinedConversionAnalysis.FromType,
+                            operandType,
+                            checkConversion: false,
+                            fromExplicitCast: false,
+                            out _);
+
+                        // PROTOTYPE(NullableReferenceTypes): Update method based on operandType
+                        // (see StaticNullChecking_FlowAnalysis.Conversions_07).
+                        var methodOpt = conversion.Method;
+                        Debug.Assert((object)methodOpt != null);
+                        Debug.Assert(methodOpt.ParameterCount == 1);
+                        var parameter = methodOpt.Parameters[0];
+
+                        // conversion "from" type -> method parameter type
+                        operandType = ClassifyAndApplyConversion(node, parameter.Type.TypeSymbol, operandType);
+
+                        // method parameter type -> method return type
+                        ReportArgumentWarnings(operandOpt, operandType, parameter);
```

Comment:
>ReportArgumentWarnings(operandOpt, operandType, parameter); [](start = 24, length = 59)

I am really confused what is going on in this method and what calls are responsible for what. For example, 
- why calling ApplyConversion above is not sufficient and we need to call ClassifyAndApplyConversion. 
- Both ApplyConversion and ClassifyAndApplyConversion change ```operandType``` what is the difference in the changes?
- Finally, we are calling ReportArgumentWarnings. What warning would it report that the ApplyConversion and ClassifyAndApplyConversion will not report?

Reply:
`ClassifyAndApplyConversion` calls `ClassifyStandardConversion`, reports nested nullability mismatches, and then calls `ApplyConversion(..., checkConversion:false, ...)` using the result of `ClassifyStandardConversion`. Top-level nullability mismatches are reported by caller as needed.

---
In reply to: [190746194](https://github.com/dotnet/roslyn/pull/26356#discussion_r190746194) [](ancestors = 190746194)